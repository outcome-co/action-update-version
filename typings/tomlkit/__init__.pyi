"""Partial stubs for tomlkit."""

from tomlkit.toml_document import TOMLDocument

def loads(string: str) -> TOMLDocument: ...
def parse(string: str) -> TOMLDocument: ...
def document() -> TOMLDocument: ...
def dumps(data: TOMLDocument, sort_keys: bool = ...) -> str: ...
