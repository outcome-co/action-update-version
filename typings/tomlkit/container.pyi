from __future__ import annotations

from typing import Any, KeysView, Union

from tomlkit.items import Item, Key

class Container(dict):  # noqa: WPS600
    def __getitem__(self, key: Union[Key, str]) -> Union[Item, Container]: ...
    def __setitem__(self, key: Union[Key, str], value: Any) -> None: ...
    def keys(self) -> KeysView[str]: ...
    def as_string(self) -> str: ...
