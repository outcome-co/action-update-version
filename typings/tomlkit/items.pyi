from __future__ import annotations

from enum import Enum
from typing import Any, ClassVar, Dict, Generator, Optional, Tuple, Type, Union

from tomlkit.container import Container

class KeyType(Enum):
    Bare: ClassVar[str]
    Basic: ClassVar[str]
    Literal: ClassVar[str]

class Key:
    key: str
    sep: str
    def __init__(
        self, k: str, t: Optional[KeyType] = ..., sep: Optional[str] = ..., dotted: bool = ..., original: Optional[str] = ...,
    ) -> None: ...
    @property
    def delimiter(self) -> str: ...
    def is_dotted(self) -> bool: ...
    def is_bare(self) -> bool: ...
    def as_string(self) -> str: ...
    def __hash__(self) -> int: ...
    def __eq__(self, other: Key) -> bool: ...
    def __str__(self) -> str: ...
    def __repr__(self) -> str: ...

class Item(object):
    """
    An item within a TOML document.
    """

    def __init__(self, trivia: Trivia) -> None: ...
    @property
    def trivia(self) -> Trivia: ...
    @property
    def discriminant(self) -> int: ...
    def as_string(self) -> str: ...
    # Helpers
    def comment(self, comment: str) -> Item: ...
    def indent(self, indent: int) -> Item: ...
    def is_boolean(self) -> bool: ...
    def is_table(self) -> bool: ...
    def is_inline_table(self) -> bool: ...
    def is_aot(self) -> bool: ...
    def _getstate(self, protocol: int = ...) -> Tuple[Trivia]: ...
    def __reduce__(self) -> Tuple[Type[Key], Tuple[Trivia]]: ...
    def __reduce_ex__(self, protocol: int) -> Tuple[Type[Key], Tuple[Trivia]]: ...

class Trivia:
    indent: str
    comment_ws: str
    comment: str
    trail: str
    def __init__(
        self,
        indent: Optional[str] = ...,
        comment_ws: Optional[str] = ...,
        comment: Optional[str] = ...,
        trail: Optional[str] = ...,
    ) -> None: ...

class Table(Item, dict):
    def __init__(
        self,
        value: Container,
        trivia: Trivia,
        is_aot_element: bool,
        is_super_table: bool = ...,
        name: Optional[str] = ...,
        display_name: Optional[str] = ...,
    ) -> None: ...
    @property
    def value(self) -> Container: ...
    @property
    def discriminant(self) -> int: ...
    def add(self, key: Union[Key, Item, str], item: Any = ...) -> Item: ...
    def append(self, key: Union[Key, str], _item: Any) -> Table: ...
    def raw_append(self, key: Union[Key, str], _item: Any) -> Table: ...
    def remove(self, key: Union[Key, str]) -> Table: ...
    def is_aot_element(self) -> bool: ...
    def is_super_table(self) -> bool: ...
    def as_string(self) -> str: ...
    # Helpers
    def indent(self, indent: int) -> Table: ...
    def keys(self) -> Generator[str, None, None]: ...
    def values(self) -> Generator[Item, None, None]: ...
    def items(self) -> Generator[Tuple[str, Item], None, None]: ...
    def update(self, other: Dict[Any, Any]) -> None: ...
    def get(self, key: Any, default: Optional[Any] = ...) -> Any: ...
    def __contains__(self, key: Union[Key, str]) -> bool: ...
    def __getitem__(self, key: Union[Key, str]) -> Item: ...
    def __setitem__(self, key: Union[Key, str], value: Any) -> None: ...
    def __delitem__(self, key: Union[Key, str]) -> None: ...
    def __repr__(self) -> str: ...
    def __str__(self) -> str: ...

class StringType(Enum):
    # Single Line Basic
    SLB = '"'
    # Multi Line Basic
    MLB = '"""'
    # Single Line Literal
    SLL = "'"
    # Multi Line Literal
    MLL = "'''"
    @property
    def unit(self) -> str: ...
    def is_basic(self) -> bool: ...
    def is_literal(self) -> bool: ...
    def is_singleline(self) -> bool: ...
    def is_multiline(self) -> bool: ...
    def toggle(self) -> StringType: ...

class String(str, Item):
    """
    A string literal.
    """

    def __new__(cls, t: StringType, value: str, original: str, trivia: Trivia) -> None: ...
    def __init__(self, t: StringType, _: str, original: str, trivia: Trivia) -> None: ...
    @property
    def discriminant(self) -> int: ...
    @property
    def value(self) -> str: ...
    def as_string(self) -> str: ...
    def __add__(self, other: str) -> String: ...
    def __sub__(self, other: str) -> String: ...
    def _new(self, result: str) -> String: ...
