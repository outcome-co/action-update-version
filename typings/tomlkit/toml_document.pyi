from tomlkit.container import Container

class TOMLDocument(Container): ...
