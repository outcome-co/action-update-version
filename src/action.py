"""Set up Commitizen.

If there is a pyproject.toml file in the directory, we'll use that to store the cz config,
otherwise we'll use .cz.toml.

The script will detect the version files to update, either pyproject.toml, package.json, or config.cfg.
"""

import json
import sys
from pathlib import Path
from typing import List, Tuple, Type, TypeVar, Union, cast

import tomlkit
import typer
from commitizen import cmd
from tomlkit.container import Container
from tomlkit.items import Item, String, Table
from tomlkit.toml_document import TOMLDocument

app = typer.Typer()
_read_mode = 'r'


@app.command()
def bump():
    # Retrieve existing tags
    run('git fetch --tags')

    # Run commitizen
    result = run('cz bump --no-verify --yes')
    updated = not any(line == 'increment detected: None' for line in result.out.splitlines())

    _set_github_output_variable('updated', str(updated).lower())

    current_version = run('cz version --project')

    _set_github_output_variable('version', current_version.out)


def run(command: str) -> cmd.Command:
    result = cmd.run(command)

    if result.return_code != 0:  # pragma: no cover
        typer.echo('Command failed!', err=True)
        typer.echo(f'stdout: {result.out}', err=True)
        typer.echo(f'stderr: {result.err}', err=True)
        sys.exit(-1)

    return result


def _set_github_output_variable(key: str, value: str):
    typer.echo(f'::set-output name={key}::{value}')


cz_toml = Path('.cz.toml')
pyproject = Path('pyproject.toml')
package_json = Path('package.json')
version_other = Path('config.cfg')

default_version = '0.1.0'

CZ_TEMPLATE = """[tool.commitizen]
name = "cz_conventional_commits"
version = ""
tag_format = "v$version"
bump_message = "chore(version): $current_version â†’ $new_version"
version_files = ["README.md:version-badge"]
\n"""


def _pyproject_exists():
    return pyproject.exists()


def _cz_toml_exists():
    return cz_toml.exists()


def _package_json_exists():
    return package_json.exists()


def _version_other_exists():
    return version_other.exists()


_already_exists_exit_code = 10
_misconfigured = 11


@app.command()
def init() -> None:  # noqa: WPS231 - too high complexity
    if _pyproject_exists():
        if is_cz_in_toml(pyproject):
            typer.echo(f'cz already configured in {pyproject}', err=True)
            sys.exit(_already_exists_exit_code)
        else:
            add_cz_config(pyproject)

    elif _cz_toml_exists():
        if not is_cz_in_toml(cz_toml):
            typer.echo(f'{cz_toml} exists, but does not contain valid config!', err=True)
            sys.exit(_misconfigured)

        typer.echo(f'cz already configured in {cz_toml}', err=True)
        sys.exit(_already_exists_exit_code)

    else:
        add_cz_config(cz_toml)


T = TypeVar('T', Table, Container, Item)


def get_from_toml(toml: Union[Container, Table], key: str, expected_type: Type[T]) -> T:
    result = toml[key]
    assert isinstance(result, expected_type)  # noqa: S101
    return result


def is_cz_in_toml(toml_file: Path) -> bool:
    """Check if the TOML file contains CZ config.

    Args:
        toml_file (Path): The TOML file to check.

    Returns:
        bool: True if the file contains CZ config
    """
    with open(toml_file, _read_mode) as file:
        toml_str = file.read()
    toml_parsed = tomlkit.loads(toml_str)

    tool = get_from_toml(toml_parsed, 'tool', Table)
    return 'commitizen' in tool.keys()


def add_cz_config(toml_file: Path) -> None:
    typer.echo(f'Adding Commitizen config to {toml_file}')

    version, version_file = get_current_version_info()
    typer.echo(f'Version set to {version} with filepath {version_file}')

    cz_config = get_cz_config(version, version_file)
    current_config = get_current_config(toml_file)

    new_config = merge_current_and_cz(current_config, cz_config)
    write_new_config(toml_file, new_config)


def get_current_version_info() -> Tuple[str, str]:
    if _pyproject_exists():
        with open(pyproject, _read_mode) as py_file:
            py_toml_str = py_file.read()
        parsed_toml = tomlkit.loads(py_toml_str)

        tool = get_from_toml(parsed_toml, 'tool', Table)
        poetry = get_from_toml(tool, 'poetry', Table)
        version = get_from_toml(poetry, 'version', String)

        return str(version), f'{pyproject}:version'

    if _package_json_exists():
        parsed_json = json.load(open(package_json))  # noqa: WPS515
        return parsed_json['version'], f'{package_json}:version'

    if _version_other_exists():
        version = open(version_other, _read_mode).read().strip()  # noqa: WPS515
        return version, str(version_other)

    with open(version_other, 'w') as version_other_handle:
        version_other_handle.write(f'{default_version}\n')

    return default_version, str(version_other)


def get_cz_config(version: str, version_file: str) -> TOMLDocument:
    cz_parsed = tomlkit.parse(CZ_TEMPLATE)

    tool = get_from_toml(cz_parsed, 'tool', Table)
    commitizen = get_from_toml(tool, 'commitizen', Table)

    commitizen['version'] = version

    version_files = cast(List[str], commitizen['version_files'])
    version_files.append((version_file))

    return cz_parsed


def get_current_config(toml_file: Path) -> TOMLDocument:
    if toml_file.exists():
        with open(toml_file, _read_mode) as f_read:
            return tomlkit.parse(f_read.read())
    return tomlkit.document()


def merge_current_and_cz(current_config: TOMLDocument, cz_config: TOMLDocument) -> TOMLDocument:
    # As current_config and cz_config are both already TOMLDocuments, the format returned by tomlkit method
    # "as_string()" will include every details needed to preserve the TOMLDocument style when merging
    merged = current_config.as_string() + cz_config.as_string()
    # Then we ensure the format is still valid by parsing the merged string in a TOMLDocument again
    return tomlkit.parse(merged)


def write_new_config(toml_file: Path, new_config: TOMLDocument) -> None:
    with open(toml_file, 'w') as f_write:
        f_write.write(tomlkit.dumps(new_config))


if __name__ == '__main__':  # pragma: no cover
    app()
