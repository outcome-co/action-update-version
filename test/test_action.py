"""Test module Setup Commitizen."""

import json
from typing import Generator, List, Optional

import pytest
import tomlkit
from click.testing import Result
from py import path
from syrupy.assertion import SnapshotAssertion
from tomlkit.items import String, Table
from typer.testing import CliRunner

from src import action

TEST_CZ_TEMPLATE = """[tool.commitizen]
name = "cz_conventional_commits"
version = "1.1.1"
tag_format = "v$version"
bump_message = "chore(version): $current_version â†’ $new_version"
version_files = ["README.md:version-badge", "pyproject.toml:version"]
\n"""


PYPROJECT_TOML = """[tool.poetry]
name = "action-update-version"
version = "0.10.8"
\n"""

runner = CliRunner(mix_stderr=False)

_already_exists_exit_code = 10
_misconfigured = 11


@pytest.fixture()
def clean_dir(tmpdir: path.local) -> Generator[path.local, None, None]:
    with tmpdir.as_cwd():
        yield tmpdir


@pytest.fixture()
def tmp_git_project(clean_dir: path.local) -> Generator[path.local, None, None]:
    with clean_dir.as_cwd():
        action.run('git init')
        yield clean_dir


@pytest.fixture()
def tmp_commitizen_project(tmp_git_project: path.local) -> Generator[path.local, None, None]:
    with tmp_git_project.as_cwd():
        tmp_commitizen_cfg_file = tmp_git_project.join('pyproject.toml')
        tmp_commitizen_cfg_file.write('[tool.commitizen]\nversion="0.1.0"\n')

        action.run('git add pyproject.toml')
        action.run('git commit -m "build: add pyproject.toml"')

        yield tmp_git_project


def github_variable_set(result: Result, key: str, value: str) -> bool:
    o = f'::set-output name={key}::{value}'
    return any(line == o for line in result.output.splitlines())


def commitizen_in_toml(path: path.local, version: str, version_files: Optional[List[str]] = None) -> bool:
    contents = path.read_text('utf-8')
    parsed = tomlkit.parse(contents)

    try:
        tool = action.get_from_toml(parsed, 'tool', Table)
        commitizen = action.get_from_toml(tool, 'commitizen', Table)
        read_version = action.get_from_toml(commitizen, 'version', String)

        if version_files:
            read_version_files = commitizen['version_files']
            assert isinstance(read_version_files, list)

            if set(version_files) != set(read_version_files):
                return False

        return version == read_version
    except Exception:
        return False


def poetry_in_toml(path: path.local) -> bool:
    contents = path.read_text('utf-8')
    parsed = tomlkit.parse(contents)

    tool = action.get_from_toml(parsed, 'tool', Table)

    return 'poetry' in tool.keys()


def check_exit_code(result: Result, expected_code: int = 0) -> None:
    if result.exit_code == expected_code:
        return

    print(result.stderr)  # noqa: T001, WPS421
    print(result.stdout)  # noqa: T001, WPS421

    if result.exception:
        print(result.exception)  # noqa: T001, WPS421

    assert result.exit_code == expected_code


@pytest.mark.usefixtures('tmp_commitizen_project')
class TestBump:
    def test_bump(self, tmp_commitizen_project: path.local):
        some_file = tmp_commitizen_project.join('some_file')
        some_file.write('content')

        action.run('git add some_file')
        action.run('git commit -m "feat: add some_file"')

        result = runner.invoke(action.app, ['bump'])

        check_exit_code(result)
        assert github_variable_set(result, 'updated', 'true')
        assert github_variable_set(result, 'version', '0.2.0')

    def test_no_bump(self, tmp_commitizen_project: path.local):
        some_file = tmp_commitizen_project.join('some_file')
        some_file.write('content')

        action.run('git add some_file')
        action.run('git commit -m "chore: add some_file"')

        result = runner.invoke(action.app, ['bump'])

        check_exit_code(result)
        assert github_variable_set(result, 'updated', 'false')
        assert github_variable_set(result, 'version', '0.1.0')


class TestInitPyProject:
    def test_already_exists(self, clean_dir: path.local):
        pyproject = clean_dir.join('pyproject.toml')
        pyproject.write(TEST_CZ_TEMPLATE)

        result = runner.invoke(action.app, ['init'])

        check_exit_code(result, _already_exists_exit_code)
        assert 'cz already configured' in result.stderr

    def test_add_config(self, clean_dir: path.local):
        pyproject = clean_dir.join('pyproject.toml')
        pyproject.write(PYPROJECT_TOML)

        result = runner.invoke(action.app, ['init'])

        check_exit_code(result)
        assert 'Adding Commitizen config to' in result.stdout
        assert commitizen_in_toml(pyproject, '0.10.8')
        assert poetry_in_toml(pyproject)


class TestInitNonPyproject:
    def test_add_config(self, clean_dir: path.local):
        result = runner.invoke(action.app, ['init'])

        check_exit_code(result)
        assert 'Adding Commitizen config to' in result.stdout

        cz_toml = clean_dir.join('.cz.toml')
        assert commitizen_in_toml(cz_toml, '0.1.0')

    def test_misconfigured(self, clean_dir: path.local):
        cz_toml = clean_dir.join('.cz.toml')
        cz_toml.write(PYPROJECT_TOML)

        result = runner.invoke(action.app, ['init'])

        check_exit_code(result, _misconfigured)
        assert 'but does not contain valid config' in result.stderr

    def test_already_configured(self, clean_dir: path.local):
        cz_toml = clean_dir.join('.cz.toml')
        cz_toml.write(TEST_CZ_TEMPLATE)

        result = runner.invoke(action.app, ['init'])

        check_exit_code(result, _already_exists_exit_code)
        assert 'cz already configured in' in result.stderr


class TestGetCurrentVersion:
    def test_get_package_json_version(self, clean_dir: path.local):
        package_json = clean_dir.join('package.json')
        package_json.write(json.dumps({'version': '0.1.2'}))
        assert action.get_current_version_info() == ('0.1.2', 'package.json:version')

    def test_get_pyproject_version(self, clean_dir: path.local):
        package_json = clean_dir.join('pyproject.toml')
        package_json.write(PYPROJECT_TOML)
        assert action.get_current_version_info() == ('0.10.8', 'pyproject.toml:version')

    def test_get_other_version(self, clean_dir: path.local):
        version_file = clean_dir.join('config.cfg')
        version_file.write('1.0.0')
        assert action.get_current_version_info() == ('1.0.0', 'config.cfg')

    def test_get_no_version(self, clean_dir: path.local):
        version_file = clean_dir.join('config.cfg')

        assert not version_file.exists()
        assert action.get_current_version_info() == ('0.1.0', 'config.cfg')
        assert version_file.exists()


class TestAddCzInfo:
    def test_correct_version_is_added(self, clean_dir: path.local):
        version_file = clean_dir.join('config.cfg')
        version_file.write('1.2.3')

        result = runner.invoke(action.app, ['init'])

        check_exit_code(result)

        cz_toml = clean_dir.join('.cz.toml')
        assert commitizen_in_toml(cz_toml, '1.2.3', ['config.cfg', 'README.md:version-badge'])

    def test_snapshot(self, clean_dir: path.local, snapshot: SnapshotAssertion):
        pyproject = clean_dir.join('pyproject.toml')
        pyproject.write(PYPROJECT_TOML)

        result = runner.invoke(action.app, ['init'])

        check_exit_code(result)

        contents = pyproject.read_text('utf-8')
        assert contents == snapshot
