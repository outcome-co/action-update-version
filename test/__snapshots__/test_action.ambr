# name: TestAddCzInfo.test_snapshot
  '
  [tool.poetry]
  name = "action-update-version"
  version = "0.10.8"
  
  [tool.commitizen]
  name = "cz_conventional_commits"
  version = "0.10.8"
  tag_format = "v$version"
  bump_message = "chore(version): $current_version â†’ $new_version"
  version_files = ["README.md:version-badge", "pyproject.toml:version"]
  
  
  '
---
